### PR TITLE
proper benchmarking

### PR DESCRIPTION
**Pull request by hexaeder**
This is a proposal on how to do proper benchmarking for future version using `PkgBenchmark.jl` and `BenchmarkTools.jl`.

I've added a [readme](https://github.com/PIK-ICoNe/NetworkDynamics.jl/tree/hw/benchmarking/benchmark) to the benchmark folder on how to run those.

In short
```
$ cd NetworkDynamics/benchmarks
$ ./run_benchmarks.jl v0.5.0    # this will take a few minutes
```
will create 3 markdownfiles with the results of the benchmarks
- the target version, i.e. the current state of the repo
- the baseline version (in this case v0.5.0) for the comparison
- a comparison with relative differences for each of the timings.

You can see those markdown files in this [gist](https://gist.github.com/hexaeder/05633f201dbfe278d3486b3ec55f0eb6).

The most importand is probably this section, where you can see the nice speedup of #80 
## Results
A ratio greater than `1.0` denotes a possible regression (marked with :x:), while a ratio less
than `1.0` denotes a possible improvement (marked with :white_check_mark:). Only significant results - results
that indicate possible regressions or improvements - are shown below (thus, an empty table means that all
benchmark results remained invariant between builds).

| ID                                                   | time ratio                   | memory ratio                 |
|------------------------------------------------------|------------------------------|------------------------------|
| `["diffusion", "ode_edge", "assemble", "10"]`        |                   0.97 (5%)  | 0.87 (1%) :white_check_mark: |
| `["diffusion", "ode_edge", "assemble", "100"]`       | 0.85 (5%) :white_check_mark: | 0.41 (1%) :white_check_mark: |
| `["diffusion", "ode_edge", "assemble", "1000"]`      | 0.32 (5%) :white_check_mark: | 0.07 (1%) :white_check_mark: |
| **`["diffusion", "ode_edge", "assemble", "10000"]`**     | 0.06 (5%) :white_check_mark: | 0.01 (1%) :white_check_mark: |
| `["diffusion", "ode_edge", "call_mt", "10"]`         | 0.93 (5%) :white_check_mark: |                   1.01 (1%)  |
| `["diffusion", "ode_edge", "call_mt", "100"]`        | 0.64 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["diffusion", "static_edge", "assemble", "10"]`     | 0.95 (5%) :white_check_mark: | 0.86 (1%) :white_check_mark: |
| `["diffusion", "static_edge", "assemble", "100"]`    | 0.84 (5%) :white_check_mark: | 0.40 (1%) :white_check_mark: |
| `["diffusion", "static_edge", "assemble", "1000"]`   | 0.31 (5%) :white_check_mark: | 0.07 (1%) :white_check_mark: |
| **`["diffusion", "static_edge", "assemble", "10000"]`** | 0.06 (5%) :white_check_mark: | 0.01 (1%) :white_check_mark: |
| `["diffusion", "static_edge", "call", "10"]`         |                1.51 (5%) :x: |                   1.00 (1%)  |
| `["diffusion", "static_edge", "call", "1000"]`       | 0.94 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["diffusion", "static_edge", "call_mt", "10"]`      | 0.82 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["diffusion", "static_edge", "call_mt", "100"]`     | 0.44 (5%) :white_check_mark: |                   0.99 (1%)  |
| `["diffusion", "static_edge", "call_mt", "1000"]`    | 0.53 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["diffusion", "static_edge", "call_mt", "10000"]`   | 0.88 (5%) :white_check_mark: |                   0.99 (1%)  |
| `["kuramoto", "heterogeneous", "assemble", "10"]`    |                   0.96 (5%)  | 0.86 (1%) :white_check_mark: |
| `["kuramoto", "heterogeneous", "assemble", "100"]`   | 0.84 (5%) :white_check_mark: | 0.40 (1%) :white_check_mark: |
| `["kuramoto", "heterogeneous", "assemble", "1000"]`  | 0.32 (5%) :white_check_mark: | 0.07 (1%) :white_check_mark: |
| **`["kuramoto", "heterogeneous", "assemble", "10000"]`** | 0.03 (5%) :white_check_mark: | 0.01 (1%) :white_check_mark: |
| `["kuramoto", "heterogeneous", "call", "10"]`        | 0.88 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["kuramoto", "heterogeneous", "call", "100"]`       | 0.85 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["kuramoto", "heterogeneous", "call", "1000"]`      | 0.85 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["kuramoto", "heterogeneous", "call", "10000"]`     | 0.83 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["kuramoto", "heterogeneous", "call_mt", "10"]`     |                1.09 (5%) :x: |                   1.00 (1%)  |
| `["kuramoto", "heterogeneous", "call_mt", "100"]`    | 0.88 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["kuramoto", "heterogeneous", "call_mt", "1000"]`   | 0.85 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["kuramoto", "heterogeneous", "call_mt", "10000"]`  | 0.87 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["kuramoto", "homogeneous", "assemble", "10"]`      |                   0.97 (5%)  | 0.87 (1%) :white_check_mark: |
| `["kuramoto", "homogeneous", "assemble", "100"]`     | 0.85 (5%) :white_check_mark: | 0.43 (1%) :white_check_mark: |
| `["kuramoto", "homogeneous", "assemble", "1000"]`    | 0.34 (5%) :white_check_mark: | 0.08 (1%) :white_check_mark: |
| **`["kuramoto", "homogeneous", "assemble", "10000"]`**   | 0.03 (5%) :white_check_mark: | 0.01 (1%) :white_check_mark: |
| `["kuramoto", "homogeneous", "call", "10"]`          | 0.75 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["kuramoto", "homogeneous", "call", "1000"]`        | 0.90 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["kuramoto", "homogeneous", "call_mt", "100"]`      |                1.22 (5%) :x: |                   1.00 (1%)  |
| `["kuramoto", "homogeneous", "call_mt", "1000"]`     | 0.66 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["kuramoto", "homogeneous", "call_mt", "10000"]`    |                   0.97 (5%)  | 0.99 (1%) :white_check_mark: |

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -4,6 +4,7 @@ authors = ["Frank Hellmann <hellmann@pik-potsdam.de>, Michael Lindner <michaelli
 version = "0.5.4"
 
 [deps]
+BenchmarkTools = "6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf"
```
Comment (lindnemi) [inline Project.toml:7]: Would it be possible to add this dependency to a local environment in the `benchmarking` folder instead?

**Review by lindnemi — COMMENTED** [review]
This is absolutely great and could be merged right away. I was wondering about a few more things though

* It would be nice to benchmark an inhomogeneous system with algebraic constraints in mass matrix form. Thus we would benchmark stiff solvers as well that should improve with the upcoming Jacobian vector product.

* Would it be possible to add these benchmarks to CI? Or should we run them on our machines manually each time we release a new Version?

* Just for clarification: If I get this right we could also benchmark against different Julia version in order catch potential regressions in Base?

File: Project.toml (hunk 1 above)
Comment (hexaeder) [reply Project.toml:7]: This is not that trivial unfortunately. Normally, `pkgbenchmark` will take the benchmark files from the specific commit you are benchmarking. I think this is not what we want here. During the process of benchmarking a different commit the script will
- copy the NetworkDynamics repo to a temp location
- copy the benchmark folder to a different temp location
- change the temp NetworkDynamics folder to the commit to we want to
- at this point, it is not clear how to reach the temp ND folder from the environment which lives inside the benchmark folder, therefore it is much simpler to run everything from the ND environment

**Comment by hexaeder** [comment]
>     * It would be nice to benchmark an inhomogeneous system with algebraic constraints in mass matrix form. Thus we would benchmark stiff solvers as well that should improve with the upcoming Jacobian vector product.

well, thats the idea. Having a common place where to add as much benchmarks as we like... right now they are also a bit redudant, i'm not sure whether we should test everything for so many values of N

>     * Would it be possible to add these benchmarks to CI? Or should we run them on our machines manually each time we release a new Version?

In theory yeah, totally possible. We can test this out at some point. For this, i think running it lokally will be sufficient. The default config will allways compare to master. I thinke once your are happy with your PR, run `./run_benchmark` on your PR branch and but the judgmentfile in the comments

>     * Just for clarification: If I get this right we could also benchmark against different Julia version in order catch potential regressions in Base?

Yes, right now i am adding more parameters to the benchmark script so you can do something like
```
$ ./run_benchmarks.jl --target directory --baseline directory --tcommand julia-nightly --bcommand julia
```
to benchmark the current directory on `julia-nightly` vs `julia`.

**Comment by lindnemi** [comment]
Unfortunately I can't run the benchmarks on my machine. I get: 
ERROR: LoadError: AssertionError: Captured sysimg looks weird? using Revise...
-J/home/micha/julia/julia-1.6.2/lib/julia/sys.so

**Comment by lindnemi** [comment]
ok great. the benchmarks make a lot of sense like that. diffusion benchmarks a star graph and kuramoto a sparse graph.

at some point we should benchmark different solver calls, especially when it comes to evaluating performance of JVPs. we could add that with the JVP PR #95.

**Pull request merged**